SourceDistributions/Developer-Sandboxes/tsWxGTUI_PyVx/Python-2x/tsWxGTUI_Py2x/ManPages/tsManPagesLibGUI/tsWxTopLevelWindow.man

Help on module tsWxTopLevelWindow:

NAME
    tsWxTopLevelWindow

FILE
    ./tsLibGUI/tsWxPkg/src/tsWxTopLevelWindow.py

DESCRIPTION
    tsWxTopLevelWindow.py - Common base class for wxDialog and
    wxFrame. It is an abstract base class meaning that you never
    work with objects of this class directly, but all of its
    methods are also applicable for the two classes above.
    
    Note that the instances of wxTopLevelWindow are managed by
    wxWidgets in the internal top level (task bar) window list.

CLASSES
    tsWxWindow.Window(tsWxEvtHandler.EvtHandler)
        TopLevelWindow
    
    class TopLevelWindow(tsWxWindow.Window)
     |  Common base class for wxDialog and wxFrame. It is an abstract base
     |  class meaning that you never work with objects of this class directly,
     |  but all of its methods are also applicable for the two classes above.
     |  
     |  Note that the instances of wxTopLevelWindow are managed by wxWidgets
     |  in the internal top level (task bar) window list.
     |  
     |  Method resolution order:
     |      TopLevelWindow
     |      tsWxWindow.Window
     |      tsWxEvtHandler.EvtHandler
     |      tsWxObject.Object
     |      __builtin__.object
     |  
     |  Methods defined here:
     |  
     |  EnableCloseButton(self, enable)
     |      Return True if close button enabled state is being changed.
     |      Return False if close button enabled state is NOT being changed.
     |  
     |  GetClosed(self)
     |      Return True if top level window has been closed.
     |      Return False if top level window has NOT been closed.
     |  
     |  GetDefaultItem(self)
     |      Get the default child of this parent, i.e. the first one
     |  
     |  GetHelpAsked(self)
     |      Return True if top level window has been asked for Help.
     |      Return False if top level window has NOT been asked for Help.
     |  
     |  GetIcon(self)
     |      TBD
     |  
     |  GetIconized(self)
     |      Return True if top level window has been iconized.
     |      Return False if top level window has NOT been iconized.
     |  
     |  GetMaximized(self)
     |      Return True if top level window has been maximuzed.
     |      Return False if top level window has NOT been maximized.
     |  
     |  GetRestoreDown(self)
     |      Return True if top level window has been restoredown.
     |      Return False if top level window has NOT been restoredown.
     |  
     |  GetTitle(self)
     |      TBD - Return title or None as appropriate.
     |  
     |  GetTmpDefaultItem(self)
     |      Return the temporary default item, which can be None.
     |  
     |  Iconize(self, iconize)
     |  
     |  IsActive(self)
     |  
     |  IsAlwaysMaximized(self)
     |  
     |  IsClosed(self)
     |  
     |  IsFullScreen(self)
     |  
     |  IsHelpAsked(self)
     |  
     |  IsIconized(self)
     |  
     |  IsMaximized(self)
     |  
     |  IsRestoreDown(self)
     |  
     |  MacGetMetalAppearance(self)
     |  
     |  MacGetUnifiedAppearance(self)
     |  
     |  MacSetMetalAppearance(self, on)
     |  
     |  Maximize(self, maximize)
     |  
     |  RequestUserAttention(self, flags)
     |  
     |  SetClosed(self, closed)
     |  
     |  SetDefaultItem(self, child)
     |      Set this child as default, return the old default.
     |  
     |  SetHelpAsked(self, helpAsked)
     |  
     |  SetIcon(self, icon)
     |  
     |  SetIconized(self, iconized)
     |  
     |  SetIcons(self, icons)
     |  
     |  SetMaximized(self, maximized)
     |  
     |  SetRestoreDown(self, restoredown)
     |  
     |  SetShape(self, region)
     |  
     |  SetTitle(self, title)
     |  
     |  SetTmpDefaultItem(self, win)
     |      Set this child as temporary default
     |  
     |  ShowFullScreen(self, show, style)
     |  
     |  __init__(self)
     |  
     |  ----------------------------------------------------------------------
     |  Data descriptors defined here:
     |  
     |  Closed
     |      Return True if top level window has been closed.
     |      Return False if top level window has NOT been closed.
     |  
     |  DefaultItem
     |      Get the default child of this parent, i.e. the first one
     |  
     |  HelpAsked
     |      Return True if top level window has been asked for Help.
     |      Return False if top level window has NOT been asked for Help.
     |  
     |  Icon
     |      TBD
     |  
     |  Iconized
     |      Return True if top level window has been iconized.
     |      Return False if top level window has NOT been iconized.
     |  
     |  Maximized
     |      Return True if top level window has been maximuzed.
     |      Return False if top level window has NOT been maximized.
     |  
     |  RestoreDown
     |      Return True if top level window has been restoredown.
     |      Return False if top level window has NOT been restoredown.
     |  
     |  Title
     |      TBD - Return title or None as appropriate.
     |  
     |  TmpDefaultItem
     |      Return the temporary default item, which can be None.
     |  
     |  ----------------------------------------------------------------------
     |  Methods inherited from tsWxWindow.Window:
     |  
     |  AcceptsFocus(self)
     |      This method may be overridden in the derived classes to return false
     |      to indicate that this control does not accept input at all
     |  
     |  AcceptsFocusFromKeyboard(self)
     |      This method may be overridden in the derived classes to return false
     |      to indicate that while this control can, in principle, have focus if
     |      the user clicks it with the mouse, it should not be included in the
     |      TAB traversal chain when using the keyboard.
     |  
     |  AcceptsFocusRecursively(self)
     |      Overridden to indicate whether this window or one of its children
     |      accepts focus.
     |  
     |  AddChild(self, child)
     |      Adds a child window. This is called automatically by window creation
     |      functions so should not be required by the application programmer.
     |  
     |  AddPendingEvent(self, event)
     |      See ProcessEvent() for more info about why you should not use
     |      this function and the reason for making this function protected
     |      in wxWindow.
     |      
     |      Reimplemented from wxEvtHandler.
     |  
     |  AdjustForLayoutDirection(self, x, width, widthTotal)
     |      Mirror coordinates for RTL layout if this window uses it and
     |      if the mirroring is not done automatically like Win32.
     |  
     |  AdjustForParentClientOriginXY(self, x, y, sizeFlags)
     |  
     |  AlwaysShowScrollbars(self)
     |      Call this function to force one or both scrollbars to be always
     |      shown, even if the window is big enough to show its entire contents
     |      without scrolling.
     |  
     |  AssociateHandle(self, handle)
     |      Associate the window with a new native handle.
     |  
     |  CacheBestSize(self, size)
     |      Sets the cached best size value.
     |  
     |  CanScroll(self, orient)
     |      Returns True if this window can have a scroll bar in this
     |      orientation.
     |  
     |  CanSetTransparent(self)
     |      Returns true if the system supports transparent windows and
     |      calling SetTransparent() may succeed.
     |      
     |      If this function returns false, transparent windows are definitely
     |      not supported by the current system.
     |      
     |      Reimplemented in wxTopLevelWindow.
     |      
     |      NOTE: On X-windows systems the X server must have the composite
     |      extension loaded, and there must be a composite manager program
     |      (such as xcompmgr) running.
     |  
     |  CaptureMouse(self)
     |      Directs all mouse input to this window.
     |      
     |      Call ReleaseMouse() to release the capture.
     |      
     |      Note that wxWindows maintains the stack of windows having captured
     |      the mouse and when the mouse is released the capture returns to the
     |      window which had had captured it previously and it is only really
     |      released if there were no previous window. In particular, this means
     |      that you must release the mouse as many times as you capture it,
     |      unless the window receives the wx.MouseCaptureLostEvent event.
     |      
     |      Any application which captures the mouse in the beginning of some
     |      operation must handle wx.MouseCaptureLostEvent and cancel this
     |      operation when it receives the event. The event handler must not
     |      recapture mouse.
     |  
     |  Center(self, direction=12)
     |      Centers the window. The parameter specifies the direction for
     |      cetering, and may be wx.HORIZONTAL, wx.VERTICAL or wx.BOTH. It may
     |      also include wx.CENTER_ON_SCREEN flag if you want to center the
     |      window on the entire screen and not on its parent window. If it is
     |      a top-level window and has no parent then it will always be centered
     |      relative to the screen.
     |  
     |  CenterOnParent(self, dir=12)
     |      Center with respect to the the parent window.
     |  
     |  CenterOnScreen(self, direction=12)
     |      Center the window on screen.
     |  
     |  Centre(self, direction=12)
     |      Centers the window. The parameter specifies the direction for
     |      cetering, and may be wx.HORIZONTAL, wx.VERTICAL or wx.BOTH. It may
     |      also include wx.CENTER_ON_SCREEN flag if you want to center the
     |      window on the entire screen and not on its parent window. If it is
     |      a top-level window and has no parent then it will always be centered
     |      relative to the screen.
     |  
     |  CentreOnParent(self, dir=12)
     |      Center with respect to the the parent window.
     |  
     |  CentreOnScreen(self, direction=12)
     |      Center the window on screen.
     |  
     |  ClearBackground(self)
     |      Clears the window by filling it with the current background colour.
     |      
     |      Does not cause an erase background event to be generated.
     |      
     |      Notice that this uses wxClientDC to draw on the window and the
     |      results of doing it while also drawing on wxPaintDC for this
     |      window are undefined. Hence this method should not be used
     |      from EVT_PAINT handlers, just use wxDC::Clear() on the wxPaintDC
     |      you already use there instead.
     |  
     |  ClientToScreen(self, pt)
     |      Converts to screen coordinates from coordinates relative to this
     |      window.
     |  
     |  ClientToScreenXY(self, x, y)
     |      Converts to screen coordinates from coordinates relative to this
     |      window.
     |  
     |  ClientToWindowSize(self, size)
     |      Converts client area size size to corresponding window size.
     |  
     |  Close(self, force=True)
     |      This function simply generates a wxCloseEvent whose handler usually
     |      tries to close the window.
     |      
     |      It does not close the window itself, however.
     |      
     |      Parameters:
     |      force   false if the windows close handler should be able to veto
     |      the destruction of this window, true if it cannot.
     |      
     |      Remarks:
     |      Close calls the close handler for the window, providing an
     |      opportunity for the window to choose whether to destroy the window.
     |      Usually it is only used with the top level windows (wxFrame and
     |      wxDialog classes) as the others are not supposed to have any special
     |      OnClose() logic. The close handler should check whether the window
     |      is being deleted forcibly, using wxCloseEvent::CanVeto, in which
     |      case it should destroy the window using wxWindow::Destroy. Note
     |      that calling Close does not guarantee that the window will be
     |      destroyed; but it provides a way to simulate a manual close of
     |      a window, which may or may not be implemented by destroying the
     |      window. The default implementation of wxDialog::OnCloseWindow
     |      does not necessarily delete the dialog, since it will simply
     |      simulate an wxID_CANCEL event which is handled by the appropriate
     |      button event handler and may do anything at all. To guarantee
     |      that the window will be destroyed, call wxWindow::Destroy instead.
     |  
     |  ConvertDialogPixels(self, pt)
     |      Converts a point or size from dialog units to pixels.
     |      
     |      For the x dimension, the dialog units are multiplied by the
     |      fixed point character width. For the y dimension, the dialog
     |      units are multiplied by the fixed point character height.
     |      
     |      Remarks:
     |      Dialog units are used for maintaining a dialogs proportions
     |      even if the font changes. You can also use these functions
     |      programmatically. A convenience macro is defined:
     |      define wxDLG_UNIT(parent, pt) parent->ConvertDialogToPixels(pt)
     |  
     |  ConvertDialogPointToPixels(self, pt)
     |      Converts a point or size from dialog units to pixels.
     |  
     |  ConvertDialogSizeToPixels(self, sz)
     |      Converts a point or size from dialog units to pixels.
     |  
     |  ConvertPixelPointToDialog(self, pt)
     |      Converts a point or size from pixel units to dialog units.
     |  
     |  ConvertPixelSizeToDialog(self, sz)
     |      Converts a point or size from pixel units to dialog units.
     |  
     |  ConvertPixelsToDialog(self, pt)
     |      Converts a point or size from pixels to dialog units.
     |      
     |      For the x dimension, the pixels are divided by the fixed point
     |      character width. For the y dimension, the pixels are divided
     |      by the fixed point character height.
     |      
     |      Remarks:
     |      Dialog units are used for maintaining a dialogs proportions even
     |      if the font changes.
     |  
     |  Create(self, parent, id=-1, pos=(-1, -1), size=(-1, -1), style=0, name='panel', pixels=True)
     |      Create the GUI part of the Window for 2-phase creation mode.
     |  
     |  DLG_PNT(self, pt)
     |      Converts a point or size from dialog units to pixels. Dialog units
     |      are used for maintaining a dialog proportions even if the font
     |      changes. For the x dimension, the dialog units are multiplied by the

Documents/NEWS.txt
#-----------------------------------------------------------
#"Time-stamp: <07/08/2015  4:48:41 AM rsg>"
#-----------------------------------------------------------

========================= NEWS.txt =========================

   +----+----+  TeamSTARS "tsWxGTUI_PyVx" Toolkit
   | ts | Wx |      with Python 2x & Python 3x based
   +----+----+         Command Line Interface (CLI)
   | G T U I |      and "Curses"-based "wxPython"-style, 
   +---------+         Graphical-Text User Interface (GUI)
 
   Get that cross-platform, pixel-mode "wxPython" feeling
   on character-mode 8-/16-color (xterm-family) & non-color
   (vt100-family) terminals and terminal emulators.

   You can find this and other plain-text files in the
   Toolkit subdirectory named:

       "./<Toolkit Recipient's Repository>/Documents".

       <Your Working Repository>
       (e.g. "tsWxGTUI_PyVx_Repository") 
         |
         |
         +-- ["Documents"]

=========================== News ===========================

The pre-alpha release of the TeamSTARS "tsWxGTUI_PyVx-0.0.0"
Toolkit is now available. It includes the following two
versions abd their associated variants (Python Distribution
Utilities ("Distutils") installable "Site-Packages" and non-
installable "Developer-Sandboxes" for enhancement and bug
fix development and experimentation without interference
with the already installed Site-Packages):

1. The TeamSTARS "tsWxGTUI_Py2x-0.0.0" Toolkit is released
   for users of the popular, high-level, cross-platform,
   second generation Python 2.x programming language which
   is widely available.

   A wiki posting by the Python Software Foundation announc-
   ed that Python 3.0 was released in 2008. The final 2.x
   version 2.7 release came out in mid-2010, with a state-
   ment of extended support for this end-of-life release.
   The 2.x branch will see no new major releases after that.

   The source code and documentation is available for down-
   load at:

      http://pypi.python.org/pypi.

   It supports development and operation of:

   a. Command Line User Interface (CLI) application
      programs implemented in Python 2.0.0-2.7.9.

   b. Graphical-style User Interface (GUI) application
      programs implemented in Python 2.6.4-2.7.9.

2. The TeamSTARS "tsWxGTUI_Py3x-0.0.0" Toolkit is released
   for users of the popular, high-level, cross-platform,
   third generation Python 3.x programming language which
   is growing in availability.

   A wiki posting by the Python Software Foundation announc-
   ed that Python 3.0 was released in 2008. 3.x is under
   active development and has already seen over five years
   of stable releases, including version 3.3 in 2012 and
   3.4 in 2014. This means that all recent standard library
   improvements, for example, are only available by default
   in Python 3.x.

   Guido van Rossum (the original creator of the Python
   language) decided to clean up Python 2.x properly,
   with less regard for backwards compatibility than is
   the case for new releases in the 2.x range. The most
   drastic improvement is the better Unicode support
   (with all text strings being Unicode by default) as
   well as saner bytes/Unicode separation.

   Besides, several aspects of the core language (such as
   print and exec being statements, integers using floor
   division) have been adjusted to be easier for newcomers
   to learn and to be more consistent with the rest of the
   language, and old cruft has been removed (for example,
   all classes are now new-style, "range()" returns a memory
   efficient iterable, not a list as in 2.x).

   The source code and documentation is available for down-
   load at:

      http://pypi.python.org/pypi.

   It supports development and operation of:

   a. Command Line User Interface (CLI) application
      programs implemented in Python 3.0.0-3.4.2.

   b. Graphical-style User Interface (GUI) application
      programs implemented in Python 3.1.5-3.4.2.

======================= End-Of-File ========================


SourceDistributions/Developer-Sandboxes/tsWxGTUI_PyVx/Python-2x/tsWxGTUI_Py2x/tsLibGUI/tsWxPkg/src/tsWxTopLevelWindow.py
#! /usr/bin/env python
# "Time-stamp: <10/25/2013  6:53:09 AM rsg>"
'''
tsWxTopLevelWindow.py - Common base class for wxDialog and
wxFrame. It is an abstract base class meaning that you never
work with objects of this class directly, but all of its
methods are also applicable for the two classes above.

Note that the instances of wxTopLevelWindow are managed by
wxWidgets in the internal top level (task bar) window list.
'''
#################################################################
#
# File: tsWxTopLevelWindow.py
#
# Purpose:
#
#    Common base class for wxDialog and wxFrame. It is an
#    abstract base class meaning that you never work with objects
#    of this class directly, but all of its methods are also
#    applicable for the two classes above.
#
#    Note that the instances of wxTopLevelWindow are managed by
#    wxWidgets in the internal top level (task bar) window list.
#
# Usage (example):
#
#    # Import
#
#    from tsWxTopLevelWindow import TopLevelWindow
#
# Requirements:
#
#    wxPython 2.8.9.2 (New wxPyDocs)/wxWidgets 2.9.2 (Ref. Man.).
#
# Capabilities:
#
#    wxPython 2.8.9.2 (New wxPyDocs)/wxWidgets 2.9.2 (Ref. Man.).
#
# Limitations:
#
#    wxPython 2.8.9.2 (New wxPyDocs)/wxWidgets 2.9.2 (Ref. Man.).
#
# Notes:
#
#    None.
#
# Methods:
#
#    wxPython 2.8.9.2 (New wxPyDocs)/wxWidgets 2.9.2 (Ref. Man.).
#
# Classes:
#
#    wxPython 2.8.9.2 (New wxPyDocs)/wxWidgets 2.9.2 (Ref. Man.).
#
# Modifications:
#
#    2010/10/03 rsg - Moved instance variables for the top-level
#                     buttons (OnClose, OnHelp, OnMaximize,
#                     OnMinimize and OnRestore) from Dialog and
#                     Frame.
#
#    2011/12/22 rsg - Added debug message for handling of
#                     tsOnCloseClick: EVT_CLOSE for [X]-Button
#                     of named Top-Level-Window.
#
#    2012/02/21 rsg Modified tsOn... to invoke
#                   tsProcessEventTables.
#
#    2012/07/28 rsg Added evt argument to each tsOn<event>.
#
#    2013/07/06 rsg Removed each tsOn<event> made obsolete
#                   by tsWxFrameButton and tsWxDialogButton.
#
#    2013/07/07 rsg Added self.ts_Closed, self.ts_HelpAsked,
#                   self.ts_Iconize, self.ts_Maximize,
#                   self.ts_RestoreDown and associated Get
#                   and Set methods and properties.
#
# ToDo:
#
#    None.
#
#################################################################

__title__     = 'tsWxTopLevelWindow'
__version__   = '1.3.0'
__date__      = '07/07/2013'
__authors__   = 'Richard S. Gordon'
__copyright__ = 'Copyright (c) 2007-2013 ' + \
                '%s.\n\t\tAll rights reserved.' % __authors__
__license__   = 'GNU General Public License, ' + \
                'Version 3, 29 June 2007'
__credits__   = '\n\n  Credits: ' + \
                '\n\n\t  tsLibGUI Import & Application Launch Features: ' + \
                '\n\t  Copyright (c) 2007-2009 Frederick A. Kier.' + \
                '\n\t\t\tAll rights reserved.' + \
                '\n\n\t  Python Curses Module API & ' + \
                'Run Time Library Features:' + \
                '\n\t  Copyright (c) 2001-2013 ' +\
                'Python Software Foundation.' + \
                '\n\t\t\tAll rights reserved.' + \
                '\n\t  PSF License Agreement for Python 2.7.3 & 3.3.0' + \
                '\n\n\t  wxWidgets (formerly wxWindows) & ' + \
                'wxPython API Features:' + \
                '\n\t  Copyright (c) 1992-2008 Julian Smart, ' + \
                'Robert Roebling,' + \
                '\n\t\t\tVadim Zeitlin and other members of the ' + \
                '\n\t\t\twxWidgets team.' + \
                '\n\t\t\tAll rights reserved.' + \
                '\n\t  wxWindows Library License' + \
                '\n\n\t  nCurses API & Run Time Library Features:' + \
                '\n\t  Copyright (c) 1998-2011 ' + \
                'Free Software Foundation, Inc.' + \
                '\n\t\t\tAll rights reserved.' + \
                '\n\t  GNU General Public License, ' + \
                'Version 3, 29 June 2007'

__line1__ = '%s, v%s (build %s)' % (__title__, __version__, __date__)
__line2__ = 'Author(s): %s' % __authors__
__line3__ = '%s' % __copyright__

if len( __credits__) == 0:
    __line4__ = '%s' % __license__
else:
    __line4__ = '%s%s' % (__license__, __credits__)

__header__ = '\n\n%s\n\n  %s\n  %s\n  %s\n' % (__line1__,
                                               __line2__,
                                               __line3__,
                                               __line4__)

mainTitleVersionDate = __line1__

#---------------------------------------------------------------------------

import tsWxGlobals as wx
from tsWxPoint import Point as wxPoint
from tsWxRect import Rect as wxRect
from tsWxSize import Size as wxSize
from tsWxWindow import Window

#---------------------------------------------------------------------------

##DEBUG = True
DEBUG = wx.Debug_GUI_Launch | \
        wx.Debug_GUI_Progress | \
        wx.Debug_GUI_Termination | \
        wx.Debug_GUI_Exceptions

##VERBOSE = True
VERBOSE = wx.Debug_GUI_Configuration

#---------------------------------------------------------------------------

class TopLevelWindow(Window):
    '''
    Common base class for wxDialog and wxFrame. It is an abstract base
    class meaning that you never work with objects of this class directly,
    but all of its methods are also applicable for the two classes above.

    Note that the instances of wxTopLevelWindow are managed by wxWidgets
    in the internal top level (task bar) window list.
    '''
    def __init__(self):
        '''
        '''
        theClass = 'TopLevelWindow'

        wx.RegisterFirstCallerClassName(self, theClass)

        Window.__init__(self, None)

        id = wx.ID_ANY

        self.tsBeginClassRegistration(theClass, id)

        self.ts_DefaultItem = None
##        self.ts_EnableCloseButton = False
        self.ts_Icons = []
##        self.thisown = None
        self.ts_Closed = False
        self.ts_HelpAsked = False
        self.ts_Iconized = False
        self.ts_Maximized = False
        self.ts_RestoreDown = False
        self.ts_ReserveTaskBarArea = wx.ThemeToUse['TaskBar']['Enable']
        self.ts_Shape = None
        self.ts_Title = None
        self.ts_TmpDefaultItem = None

        self.tsEndClassRegistration(theClass)

    #-----------------------------------------------------------------------

##    def CenterOnScreen(self, dir=wx.BOTH):
##        '''
##        Center the window on screen.
##        '''
##        self.CenterOnScreen(direction=dir)

    #-----------------------------------------------------------------------
 
##    def CentreOnScreen(self, dir=wx.BOTH):
##        '''
##        Center the window on screen.
##        '''
##        self.CenterOnScreen(direction=dir)

    #-----------------------------------------------------------------------

    def EnableCloseButton(self, enable):
        '''
        Return True if close button enabled state is being changed.
        Return False if close button enabled state is NOT being changed.
        '''
        if enable:

            if self.ts_EnableCloseButton == enable:
                return (False)
            else:
                self.ts_EnableCloseButton = enable
            return (True)

        else:

            if self.ts_EnableCloseButton == enable:
                return (False)
            else:
                self.ts_EnableCloseButton = enable
            return (True)

    #-----------------------------------------------------------------------

    def GetClosed(self):
        '''
        Return True if top level window has been closed.
        Return False if top level window has NOT been closed.
        '''
        if self.ts_Closed:
            return (True)
        else:
            return (False)

    #-----------------------------------------------------------------------

    def GetDefaultItem(self):
        '''
        Get the default child of this parent, i.e. the first one
        '''
        if self.Children is None:
            return (None)
        else:
            return self.Children[0]

    #-----------------------------------------------------------------------

    def GetHelpAsked(self):
        '''
        Return True if top level window has been asked for Help.
        Return False if top level window has NOT been asked for Help.
        '''
        if self.ts_HelpAsked:
            return (True)
        else:
            return (False)

    #-----------------------------------------------------------------------

    def GetIcon(self):
        '''
        TBD
        '''
        if self.ts_Icons == []:
            return (None)
        else:
            return (self.ts_Icons[0])

    #-----------------------------------------------------------------------

    def GetIconized(self):
        '''
        Return True if top level window has been iconized.
        Return False if top level window has NOT been iconized.
        '''
        if self.ts_Iconized:
            return (True)
        else:
            return (False)

    #-----------------------------------------------------------------------

    def GetMaximized(self):
        '''
        Return True if top level window has been maximuzed.
        Return False if top level window has NOT been maximized.
        '''
        if self.ts_Maximized:
            return (True)
        else:
            return (False)

    #-----------------------------------------------------------------------

    def GetRestoreDown(self):
        '''
        Return True if top level window has been restoredown.
        Return False if top level window has NOT been restoredown.
        '''
        if self.ts_RestoreDown:
            return (True)
        else:
            return (False)

    #-----------------------------------------------------------------------

    def GetTitle(self):
        '''
        TBD - Return title or None as appropriate.
        '''
        try:
            return (self.ts_Title)
        except AttributeError:
            return (None)

    #-----------------------------------------------------------------------

    def GetTmpDefaultItem(self):
        '''
        Return the temporary default item, which can be None.
        '''
        return (self.ts_TmpDefaultItem)

    #-----------------------------------------------------------------------

    def Iconize(self, iconize):
        '''
        '''
        pass

    #-----------------------------------------------------------------------

    def IsActive(self):
        '''
        '''
        return (False)

    #-----------------------------------------------------------------------

    def IsAlwaysMaximized(self):
        '''
        '''
        return (False)

    #-----------------------------------------------------------------------

    def IsClosed(self):
        '''
        '''
        return (self.ts_Closed)

    #-----------------------------------------------------------------------

    def IsFullScreen(self):
        '''
        '''
        return (False)

    #-----------------------------------------------------------------------

    def IsHelpAsked(self):
        '''
        '''
        return (self.ts_HelpAsked)

    #-----------------------------------------------------------------------

    def IsIconized(self):
        '''
        '''
        return (self.ts_Iconized)

    #-----------------------------------------------------------------------

    def IsMaximized(self):
        '''
        '''
        return (self.ts_Maximized)

    #-----------------------------------------------------------------------

    def IsRestoreDown(self):
        '''
        '''
        return (self.ts_RestoreDown)

    #-----------------------------------------------------------------------

    def MacGetMetalAppearance(self):
        '''
        '''
        pass

    #-----------------------------------------------------------------------

    def MacGetUnifiedAppearance(self):
        '''
        '''
        pass

    #-----------------------------------------------------------------------

    def MacSetMetalAppearance(self, on):
        '''
        '''
        pass

    #-----------------------------------------------------------------------

    def Maximize(self, maximize):
        '''
        '''
        self.ts_Maximize = maximize

    #-----------------------------------------------------------------------

    def RequestUserAttention(self, flags):
        '''
        '''
        pass

    #-----------------------------------------------------------------------

    def RestoreDown(self, restore):
        '''
        '''
        self.ts_RestoreDown = restore

    #-----------------------------------------------------------------------

    def SetClosed(self, closed):
        '''
        '''
        if closed:
            self.ts_Closed = True
            self.ts_HelpAsked = False
            self.ts_Iconized = False
            self.ts_Maximized = False
            self.ts_RestoreDown = False
        else:
            self.ts_Closed = False

    #-----------------------------------------------------------------------

    def SetDefaultItem(self, child):
        '''
        Set this child as default, return the old default.
        '''
        oldDefault = self.DefaultItem
        self.DefaultItem = child
        return (oldDefault)

    #-----------------------------------------------------------------------

    def SetHelpAsked(self, helpAsked):
        '''
        '''
        if (not self.ts_Closed) and helpAsked:
            self.ts_HelpAsked = True
        else:
            self.ts_HelpAsked = False

    #-----------------------------------------------------------------------

    def SetIcon(self, icon):
        '''
        '''
        self.ts_Icons.append(icon)

    #-----------------------------------------------------------------------

    def SetIconized(self, iconized):
        '''
        '''
        if (not self.ts_Closed) and iconized:
            self.ts_Iconized = True
        else:
            self.ts_Iconized = False

    #-----------------------------------------------------------------------

    def SetIcons(self, icons):
        '''
        '''
        for icon in icons:
            self.ts_Icons.append(icon)

    #-----------------------------------------------------------------------

    def SetMaximized(self, maximized):
        '''
        '''
        if (not self.ts_Closed) and maximized:
            self.ts_Maximized = True
        else:
            self.ts_Maximized = False

    #-----------------------------------------------------------------------

    def SetRestoreDown(self, restoredown):
        '''
        '''
        if (not self.ts_Closed) and restoredown:
            self.ts_RestoreDown = True
        else:
            self.ts_RestoreDown = False

    #-----------------------------------------------------------------------

    def SetShape(self, region):
        '''
        '''
        self.ts_Shape = region

    #-----------------------------------------------------------------------

    def SetTitle(self, title):
        '''
        '''
        self.ts_Title = title

    #-----------------------------------------------------------------------

    def SetTmpDefaultItem(self, win):
        '''
        Set this child as temporary default
        '''
        self.ts_TmpDefaultItem = win

    #-----------------------------------------------------------------------

    def ShowFullScreen(self, show, style):
        '''
        '''
        # TBD - Under Construction.
        # Extent resize of the top level window to include its children.
        theDisplay = self.Display
 
        theDisplayPos = wxPoint(theDisplay.ClientArea.x,
                                theDisplay.ClientArea.y)

        theDisplaySize = wxSize(theDisplay.ClientArea.width,
                                theDisplay.ClientArea.height)

        self.logger.debug(
            '      ShowFullScreen (screen): pos=%s; size=%s.' % (
                str(theDisplayPos), str(theDisplaySize)))

        self.Rect = wxRect(theDisplayPos.x,
                           theDisplayPos.y,
                           theDisplaySize.width,
                           theDisplaySize.height)

        self.logger.debug(
            '      Need Event to ShowFullScreen (rect): %s.' % (
                str(self.Rect)))

        return (show)

    #-----------------------------------------------------------------------

    DefaultItem = property(GetDefaultItem, SetDefaultItem)
    Closed = property(GetClosed, SetClosed)
    HelpAsked = property(GetHelpAsked, SetHelpAsked)
    Icon = property(GetIcon, SetIcon)
    Iconized = property(GetIconized, SetIconized)
    Maximized = property(GetMaximized, SetMaximized)
    RestoreDown = property(GetRestoreDown, SetRestoreDown)
##    thisown: The membership flag
    Title = property(GetTitle, SetTitle)
    TmpDefaultItem = property(GetTmpDefaultItem, SetTmpDefaultItem)

#---------------------------------------------------------------------------

if __name__ == '__main__':

    print(__header__)
